x-pack/plugins/infra/public/hooks/use_lens_attributes.ts

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useEffect, useMemo, useState } from 'react';
+import type { DataView } from '@kbn/data-views-plugin/public';
+import { Filter, Query, TimeRange } from '@kbn/es-query';
+import { useKibana } from '@kbn/kibana-react-plugin/public';
+import { FormulaPublicApi } from '@kbn/lens-plugin/public';
+import { ActionExecutionContext } from '@kbn/ui-actions-plugin/public';
+import { i18n } from '@kbn/i18n';
+import { InfraClientSetupDeps, LensAttributes, LensOptions } from '../types';
+import {
+  buildLensAttributes,
+  HostLensAttributesTypes,
+  hostMetricsLensAttributes,
+} from '../common/visualizations';
+
+interface UseLensAttributesParams {
+  type: HostLensAttributesTypes;
+  dataView: DataView | undefined;
+  options?: LensOptions;
+}
+
+export const useLensAttributes = ({
+  type,
+  dataView,
+  options = {
+    breakdownSize: 10,
+  },
+}: UseLensAttributesParams) => {
+  const {
+    services: { lens },
+  } = useKibana<InfraClientSetupDeps>();
+  const { navigateToPrefilledEditor } = lens;
+  const [formulaAPI, setFormulaAPI] = useState<FormulaPublicApi>();
+
+  useEffect(() => {
+    const getFormulaApi = async () => {
+      const { formula } = await lens.stateHelperApi();
+      return setFormulaAPI(formula);
+    };
+    getFormulaApi();
+  }, [lens]);
+
+  const attributes: LensAttributes | null = useMemo(() => {
+    if (!dataView || !formulaAPI) {
+      return null;
+    }
+
+    const VisualizationClass = hostMetricsLensAttributes[type];
+    const visualizationAttributes = buildLensAttributes(
+      new VisualizationClass(dataView, options, formulaAPI)
+    );
+
+    return visualizationAttributes;
+  }, [dataView, formulaAPI, options, type]);
+
+  const injectData = useCallback(
```

Comment:
**thought:** Both `injectData` and `getExtraActions` are later used only for computing something inside the `MetricChart` component, are these memorizations necessary? JS is really quick at declaring functions and since these handlers are not passed down to any component and so memorized to prevent a rerender, I think it could just add some extra work for the hook to memorize them.